### DISCUSSION TITLE
Can/Should we stop firebase inserting the gtag script and instead defer to Google Tag Manager?

### DISCUSSION THREAD

**Discussion by Francesco-Lanciana** [firebase/firebase-js-sdk#4927] [Q&A]
Hello!

I've found that Firebase SDK will insert a gtag script into the page to enable logging events using `firebase.analytics()`. It will even do this if you don't include a measurement ID because it actually queries your apps config dynamically (which includes a measurement ID).

This is a pain when it comes to trying to use Google Tag Manager. Ideally I want everything to be done via Tag Manager and I'm worried that having both on the page will double up on events being counted (especially because a number of events are [collected by default](https://support.google.com/analytics/answer/9234069)). 

I can see this going one of three ways:
1. It's actually OK to use Firebase (and therefore gtag) and Google Tag Manager on the same page - it doesn't double count. We just shouldn't log any events via `firebase.analytics()` but instead with GTM.
2. We leave Firebase to inject gtag and therefore don't use a GA4 configuration tag. Problem here is that GA4 event tags won't you to reference a GA4 configuration tag, otherwise you have to set a manual ID which would be really cumbersome.
3. We somehow stop Firebase from injecting the gtag and defer entirely to GTM

Option number 3 seems to be the simplest to reason about however I'm not sure if the Firebase SDK actually provides a way to turn off analytics.

Would love any tips or tricks to deal with this!

**Comment by Feiyang1** [comment]
I'm not clear how GTM and gtag interact with each other. @hsubox76, would you know?

Just to understand your use case - It seems you have linked your Firebase web app to GA, and obviously you want to log analytics events to this measurementId. Do you use any Firebase growth products like Remote config or Cloud messaging?

If you do, you have to use `firebase.analytics()` in order to build an audience that Firebase can use to target. `firebase.analytics()` attaches a special firebase id (different for every browser instance) in every event you send, and Firebase targeting relies on this special id to identify clients.
If you don't, then you can use GTM entirely without even including `firebase/analytics`, but I don't recommend it because if you decided to use any growth products from Firebase in the future, you won't have any audience, and you will have to switch to use `firebase.analytics()` to build audiences from scratch.

**Reply by Francesco-Lanciana** [reply]
Interesting, I thought Firebase Analytics was essentially just a UI layer in front of Google Analytics, so as long as you sent events to the correct GA data stream then Firebase Analytics would pick it up. And `firebase.analytics()` was just a convenience so you didn't have to install gtag and use gtag yourself. So if I understand correctly, the only extra thing `firebase.analytics()` is doing compared to gtag is that it attaches this special firebase ID?

We don't currently use Remote config or Cloud messaging, however it's possible we may use them in the future so I would like to keep that door open.

But it sucks that we would be forced to hard code all of our events and we wouldn't be able to use something like GTM that lets us add more events without changing the source code...

Keen to hear if you know of a sane way of exclusively using GTM to manage events. Perhaps we could add this special firebase ID to the datalayer and include it in any events created through GTM?

Also if we did try to go this route, after some more experimenting it seems like as long as you don't call firebase.analytics it won't load in gtag, so all we have to figure out is how to use this firebase ID in GTM

**Reply by hsubox76** [reply]
You theoretically should be able to use GTM and Firebase Analytics on the same page and I did test it on launch, but I haven't tested it extensively since, so you may turn up issues. There was one reported issue I remember but the person never replied to us when we asked for further information, so I wasn't able to look further into it: https://github.com/firebase/firebase-js-sdk/issues/4856.

Yes, if you don't initialize `firebase.analytics()`, no Firebase Analytics code will run, including gtag script injection. (If you don't want Firebase Analytics I also suggest not importing `firebase/analytics` in the first place, hopefully you are importing components individually instead of the all-in-one `import firebase from 'firebase'`.)

As mentioned above you could go ahead and just use GTM with the measurement ID if you don't want the special Firebase ID that's needed for audiences and targeting.

**Reply by Francesco-Lanciana** [reply]
@hsubox76 wouldn't you get doubling up of default events (e.g. pageview) if you had both GTM and Firebase Analytics on the same page?

In regards to this Firebase ID I have found where it is injected into the dataLayer in the Firebase SDK ([Link](https://github.com/firebase/firebase-js-sdk/blob/2a5039ee3242fb4109da9dee36ac978d78519334/packages/analytics/src/initialize-ids.ts#L143)), however I don't see where it's being attached to every event. I'm also a little confused about why firebase analytics pushes config to the dataLayer if all it's using is gtag, since I was under the impression that only GTM uses the datalayer.

**Comment by hsubox76** [comment]
In that line, it's added in a gtag config call. Properties added to a measurement ID in gtag config should be sent with every subsequent event to the same measurement ID.

As far as duplicate events: when a user puts gtag on their page, the initial page view is only triggered by the `gtag('config')` call at the end of the snippet, so if gtag users (not your case I know, see next paragraph) remove that line, it'll allow Firebase to do the initial config call (which contains the Firebase ID) and avoid doubling up. I actually do need to change [our documentation](https://firebase.google.com/docs/analytics/get-started?platform=web#firebase-gtag) on this. It currently says it's the measurementId in the script url which is incorrect.

I'm not clear what the equivalent is in GTM but [looking around](https://support.google.com/tagmanager/answer/6107124?hl=en) it seems like this is handled in the backend console by creating a tag with a "Page View" track type and setting it to trigger on "All Pages". If you disable that, then it shouldn't send duplicate initial page view events with Firebase Analytics. Let me know if I'm incorrect. Also let me know if I'm correct, and I'll also add this to the documentation (we don't have a section for GTM users right now).

Edit: Also if you want to continue to use GTM page views and disable page_view logging from Firebase Analytics instead, we're working on making that functionality available in the new modular API (v9), by letting you provide an initial config option where you can set the Google Analytics param `'send_page_view': false` and any other Google Analytics config param you want.
